### PR TITLE
Rework `ddev list` command

### PR DESCRIPTION
**Pull request by beeradb**
## The Problem:
The current `ddev list` command is some of the oldest code in ddev, and not particularly well written as it doesn't take advantage of a lot of the newer structs we've introduced. It's also lagging behind on things like the router update.

## The Fix:
Let's rework this code to use the `FindContainersByLabel` function and the app interface, while updating the output to be current.

Here's some example output (from during a test run) with these changes applied:
```
1 local site found.
NAME   	TYPE   	LOCATION               	URL                      	DATABASE URL           	STATUS
drupal8	drupal8	~/Projects/ddev/drupal8	http://drupal8.ddev.local	drupal8.ddev.local:3306	running
```

Describe output:
```
NAME   	TYPE   	LOCATION               	URL                      	DATABASE URL           	STATUS
drupal8	drupal8	~/Projects/ddev/drupal8	http://drupal8.ddev.local	drupal8.ddev.local:3306	running

MySQL Credentials
-----------------
Username:       	root
Password:       	root
Database name:  	data
Connection Info:	drupal8.ddev.local:3306

Other Services
--------------
MailHog:   	http://drupal8.ddev.local:8025
phpMyAdmin:	http://drupal8.ddev.local:8036
```

## The Test:
Simply spin up a site (or set of sites) and use `ddev list` as you normally would.

**NOTE:** this introduces a new `com.ddev.platform` container label which is used in sorting. You'll need to clear out existing docker-compose.yaml files for the list to work correctly 

## Automation Overview:
This keeps / updates the old binary based ddev list test. It also introduces unit tests for the `platform.GetApps()` function which does the heavy lifting for the `ddev list` command

## Related Issue Link(s):
#153 is the issues originally opened for this rework.

### CODE REVIEW

File: pkg/ddevapp/templates.go [hunk 1]
```diff
@@ -52,6 +54,11 @@ services:
     container_name: local-${DDEV_SITENAME}-dba
     image: $DDEV_DBAIMAGE
     restart: always
+    labels:
+      com.ddev.site-name: ${DDEV_SITENAME}
+      com.ddev.container-type: db
```
Comment (tannerjfco) [inline pkg/ddevapp/templates.go]: s/db/dba

File: pkg/ddevapp/templates.go [hunk 2]
```diff
@@ -52,6 +54,11 @@ services:
...
+      com.ddev.platform: {{ .plugin }}
+      com.ddev.app-type: {{ .appType }}
```
Comment (tannerjfco) [inline pkg/ddevapp/templates.go]: May as well have docroot/approot here for consistency.

File: README.md [hunk 3]
```diff
@@ -89,8 +89,8 @@ To see a list of your current sites you can use `ddev list`.
 ```
 ➜  ddev list
 1 local site found.
-NAME   	TYPE   	URL                      	DATABASE URL   	STATUS
-drupal8	drupal8	http://drupal8.ddev.local	127.0.0.1:32852	running
+NAME   	TYPE   	LOCATION                        URL                      	DATABASE URL
+drupal8	drupal8	/Users/username/drupal8/docroot	http://drupal8.ddev.local	drupal8.ddev.local:3306
```
Comment (tannerjfco) [inline README.md]: We discussed in #153 shortening homedir if we detect it to help keep the width short. It's not critical by any means, but it would be nice to have.

**Comment by beeradb** [comment]
I updated this a bit to fix @tannerjfco's feedback. 

I also made the following changes/improvements:
- Consolidated a bit of logic between this PR and the `ddev describe` functionality. now both list and describe share the same header row (and share code). 
- Improved the `util.GetContainerHealth()` function to return exited if containers are not running.
- Added a status column to the listing
- Introduced a new `SiteStatus()` function on the app struct to return site status as a string.

These changes are probably worth another once-over by @tannerjfco or @rfay.

This probably requires signoff from @rickmanelius since it both changes describe output slightly and introduces new list functionality which he has not yet approved. Although it's a clear improvement, he may still have input or request changes. Going ahead and requesting a review from him now.

**Comment by rfay** [comment]
Since you're working on GetContainerHealth() you might consider handling

```
35c8d15b54a8        drud/nginx-php-fpm7-local:v0.3.5-11-gfc55df1   "/start.sh"              5 hours ago         Restarting (1) Less than a second ago
```

Otherwise I can do a followup issue. Just noticed you were improving it anyway @beeradb

**Comment by rfay** [comment]
Here's an example, created by having a broken /start.sh in the web container:

`$ docker inspect --format='{{json .State.Health}}' 35c8d15b54a8{"Status":"starting","FailingStreak":0,"Log":[{"Start":"2017-04-21T14:34:45.014866116Z","End":"2017-04-21T14:34:45.015235008Z","ExitCode":-1,"Output":"No such exec instance '09286a8cfb03a624311bb1013c25b25bd8350a3da175c717ffc97b7bf758ff11' found in daemon"}]}
rfay@Randy-Air:~/go/src/github.com/drud/ddev$ docker inspect --format='{{json .State.Health}}' 35c8d15b54a8{"Status":"starting","FailingStreak":2,"Log":[{"Start":"2017-04-21T14:35:37.080357521Z","End":"2017-04-21T14:35:37.080844782Z","ExitCode":-1,"Output":"No such exec instance '6774d3fdbd65e00eb2bcced7d49087329d308cb71f2a8d60cb7e4a6e57877a5f' found in daemon"},{"Start":"2017-04-21T14:36:07.08196898Z","End":"2017-04-21T14:36:07.082467329Z","ExitCode":-1,"Output":"No such exec instance '7ef8ab8711511d533dadb4e35ae51f666e8fab078f4afb39e4e3a0b868d91c04' found in daemon"},{"Start":"2017-04-21T14:36:37.086971593Z","End":"2017-04-21T14:36:37.087718567Z","ExitCode":-1,"Output":"No such exec instance '7bb0eed2dd2d46582982035b0fbf2079dc0d957faf148a6f2b11f49bd5229b44' found in daemon"},{"Start":"2017-04-21T14:37:20.221239816Z","End":"2017-04-21T14:37:20.222146615Z","ExitCode":-1,"Output":"No such exec instance '46cf751f7ae38344ed02d3a00a5f7227a6661cef225c427a81e1541d9fe4d411' found in daemon"},{"Start":"2017-04-21T14:37:50.223660781Z","End":"2017-04-21T14:37:50.224029674Z","ExitCode":-1,"Output":"No such exec instance 'ff2b38bf88912e36b8ba1cacd60a3e6e8ffed9f80718867769015563b2f0e631' found in daemon"}]}`

**Comment by rfay** [comment]
My first experience, in a terminal window that's half the width of my screen, comes out like this, which is a lot of wrapping and hard to read. Could we drop the database info out of there?

6__default__bash__and_slack_-_drud

**Comment by rfay** [comment]
Suggestions for compactifying - remove db portion (leave it to info) and tighten up the columns.

6__default__bash__and_slack_-_drud

File: pkg/plugins/platform/local.go [hunk 4]
```diff
@@ -79,19 +80,17 @@ func (l *LocalApp) Describe() (string, error) {
...
+	output = fmt.Sprint(appTable)
 
 	output = output + "\n\nMySQL Credentials\n-----------------\n"
 	dbTable := uitable.New()
 	dbTable.MaxColWidth = maxWidth
 	dbTable.AddRow("Username:", "root")
 	dbTable.AddRow("Password:", "root")
 	dbTable.AddRow("Database name:", "data")
-	dbTable.AddRow("Connection Info:", l.HostName()+":3306")
+	dbTable.AddRow("Connection Info:", l.HostName()+":"+appports.GetPort("db"))
```
Comment (rfay) [inline pkg/plugins/platform/local.go]: Probably should be in a different PR right but since reworking getPort() in this one goes here. I'm glad to see improvements like this, and sometimes it just works best to piggy-back them.

File: pkg/plugins/platform/local.go [hunk 5]
```diff
@@ -186,6 +185,24 @@ func (l *LocalApp) ImportDB(imPath string) error {
 	return nil
 }
 
+// SiteStatus returns the current status of an application based on the web container.
+func (l *LocalApp) SiteStatus() string {
+	webContainer, err := l.FindContainerByType("web")
+	if err != nil {
+		return "not found"
```
Comment (rfay) [inline pkg/plugins/platform/local.go]: We should at least emit the error found here I'd say, otherwise we're just hiding the result with "not found". I'd be OK with a log.

File: pkg/plugins/platform/local.go [hunk 6]
```diff
@@ -186,6 +185,24 @@ func (l *LocalApp) ImportDB(imPath string) error {
...
+	}
+
+	status := util.GetContainerHealth(webContainer)
+	if status == "exited" {
+		return "stopped"
+	}
+	if status == "healthy" {
+		return "running"
+	}
```
Comment (rfay) [inline pkg/plugins/platform/local.go]: Don't we need restarting or other options here? (If we get to 3 options it's nicer to do a switch statement.)  

There are a couple of problems with this SiteStatus() function that we already know about:
* It doesn't take into account other containers (just web)
* It's an explicit translation approach from container->site, which may not be adequate.

File: pkg/plugins/platform/types.go [hunk 7]
```diff
@@ -1,5 +1,7 @@
 package platform
 
+import "log"
```
Comment (rfay) [inline pkg/plugins/platform/types.go]: Here I thought you always insisted on `	log "github.com/Sirupsen/logrus"
` :)
I might do a PR that just globally switches that out wherever we have log.

File: pkg/plugins/platform/utils.go [hunk 8]
```diff
@@ -39,50 +40,86 @@ func PrepLocalSiteDirs(base string) error {
...
+			"com.ddev.container-type": "web",
+		}
+		sites, err := util.FindContainersByLabels(labels)
 
-	for _, container := range containers {
-		for _, containerName := range container.Names {
-			for pName, plugin := range PluginMap {
-				if strings.HasPrefix(containerName[1:], plugin.ContainerPrefix()) {
-					util.ProcessContainer(apps[pName], pName, containerName[1:], container)
+		if err == nil {
```
Comment (rfay) [inline pkg/plugins/platform/utils.go]: Needs handling of err here, and I'd rather see explicit handling of err *first* and then the non-error case. Too hard to follow down otherwise.

File: pkg/plugins/platform/local_test.go [hunk 9]
```diff
@@ -133,6 +133,22 @@ func TestLocalStart(t *testing.T) {
...
+func TestGetApps(t *testing.T) {
+	assert := assert.New(t)
+	apps := GetApps()
+	assert.Equal(len(apps["local"]), len(TestSites))
+
+	for _, site := range TestSites {
+		var found bool
+		for _, siteInList := range apps["local"] {
+			if site.Name == siteInList.GetName() {
+				found = true
```
Comment (rfay) [inline pkg/plugins/platform/local_test.go]: Might as well break here.

File: pkg/plugins/platform/utils.go [hunk 10]
```diff
@@ -39,50 +40,86 @@ func PrepLocalSiteDirs(base string) error {
...
+		fmt.Sprintf("%s:%s", site.HostName(), appports.GetPort("db")),
+		site.SiteStatus(),
+	)
 }
 
 // EnsureDockerRouter ensures the router is running.
 func EnsureDockerRouter() {
-	userHome := homedir.Get()
+	userHome, err := homedir.Dir()
+	util.CheckErr(err)
```
Comment (rfay) [inline pkg/plugins/platform/utils.go]: homedir.Dir() might legitimately return an err right?  Maybe no home directory? Or user has no home? Or OS not supported by library?

File: pkg/util/dockerutils.go [hunk 11]
```diff
@@ -193,6 +168,9 @@ func ContainerWait(timeout time.Duration, labels map[string]string) error {
 				return errors.New("failed to query container")
 			}
 			status := GetContainerHealth(container)
+			if status == "restarting" {
+				return fmt.Errorf("container %s: detected container restart; invalid configuration or container. consider using `docker logs %s` to debug", container.Names[0], container.ID)
```
Comment (rfay) [inline pkg/util/dockerutils.go]: THanks!

File: pkg/plugins/platform/types.go (hunk 7 above)
Comment (beeradb) [inline pkg/plugins/platform/types.go]: I don't actually insist on `logrus`. If it's come across that way I apologize! I do however encourage people to use it all over if they need any of the advanced functionality not in the standard log package (like `.Info()`, `.Debug()`, etc.). Typically I'll switch to logrus without hesitation if I require any of that.

File: pkg/plugins/platform/utils.go (hunk 8 above)
Comment (beeradb) [inline pkg/plugins/platform/utils.go]: FindContainersbyLabel() returns an error if no containers are found. In this case, that's a valid situation (a plugin type exists but no sites are running for it). The list functionality accurately reports "no running sites" in this case.

Comment (rfay) [inline pkg/plugins/platform/utils.go]: I'm good with that if it's the only kind of error that can return.

File: pkg/plugins/platform/utils.go (hunk 10 above)
Comment (beeradb) [inline pkg/plugins/platform/utils.go]: I swapped this package out specifically because it claims to be cross-platform between POSIX and windows. 

It's certainly possible that a user might not have a home directory, but I'm not sure we could more gracefully handle the error in that case (at this point, at least) other than just fataling. The purpose of this function is to write the docker-compose.yaml for the router, which it does under `~/.ddev/router-compose.yaml`. We should maybe open a ticket to be more resilient than that, but I'm not sure we have a better option than to just fatal at this juncture.

Comment (beeradb) [inline pkg/plugins/platform/utils.go]: I'll also update and provide a more descriptive error message here, which should improve things for now.

File: pkg/plugins/platform/local.go (hunk 5 above)
Comment (beeradb) [inline pkg/plugins/platform/local.go]: Only two outcomes (that I'm aware of) could cause `webContainer` to be empty.

1. Docker is not started, a log.fatal will be thrown by `FindContainerByType` with that as an error message.
2. The container wasn't found. In the context of list, this would be very curious indeed, since we build the actual list by looking at running/stopped docker containers. 

I think in this case we're probably OK as-is, since it really really should never happen anyway, unless a container is docker-rm'd between the `GetApps()` call and the `SiteStatus()` call in the listing. If we run into situations in the future where this isn't adequate, I'm all for changing, but it's hard to know what that looks like without context I think.

**Comment by beeradb** [comment]
I removed DB info from the list output, and reduced the width in `tabwriter` from 200 characters to 140. The resulting list is quite compact on my screen.

I believe we're in good shape with this, but I'd still love a thumbsup from @rickmanelius

**Comment by rfay** [comment]
Thanks for the work on this! 

This is what it looks like now on a half-screen iterm window:

6__default__bash__and_slack_-_drud

Could we lose 'TYPE' ? Seems like people already know what type their site is?

**Comment by beeradb** [comment]
@rfay we could, but at this point, the output is shared between list and describe. It'd be awesome to keep code consolidated if possible. If we update it here, we'd need to find a way to work it into the larger describe output 

I feel like we (i.e. rick) should decide how small we want to go there.

Using the following output as an example:
```
1 local site found.
NAME   	TYPE   	LOCATION               	URL                      	STATUS
drupal8	drupal8	~/Projects/ddev/drupal8	http://drupal8.ddev.local	running
```

When displayed in a half screen iterm on my 15 inch MacBook pro monitor has this plenty of room left over (~25 additional characters).  It takes up maybe 1/8th width (at most) of my 28-inch external monitor at the default font size. 

It's hard to tell from a screenshot, but the output for your screenshot looks to be about 85 characters wide. If we reduce it beyond that, it feels like we should probably just stick to a strict 80 character width? 

tl;dr: screen size/width is such a tricky beast 💩  For me, this output is actually already smaller than I would prefer due to my normal working environments (larger laptop and/or external monitor). Understandably, for you, typically working on a 13 inch macbook air (I think?) we're pushing width limits.

**Comment by tannerjfco** [comment]
@beeradb Could we move "Status" next to name? That would ensure name and status are easily seen regardless of wrapping, and if wrapping does occur, it will be on either location or URL, which has the most potential to cause wrapping anyways.

**Comment by rickmanelius** [comment]
I may be missing something, but I'm experiencing a possible regression. I've pulled down the latest commit on the drud/ddev at beeradb/improved-list and then ran `make clean` then `make darwin`. I restarted my machine and then then confirmed sites were up by running `ddev start` and confirming active containers with `docker ps`. When I run `ddev list`, I only get "There are no running ddev applications." @beeradb

**Comment by beeradb** [comment]
@rickmanelius this introduces a new container label to help track platform. Clearing out your old `docker-compose.yaml` file and re-running start should allow list to work as expected. I apologize for not having that in the testing instructions

**Comment by rickmanelius** [comment]
Feedback:

- Not to incite a tab/space holy war, but is there any objection to putting two spaces between each column versus a tab? Seems like it would be more economical do use a consistent spacing whereas the tabs range from 1-4 spaces per instance.
- Any objection with dropping http://? macOS users can still control+click into it and it would save 7 characters each and every time.
- On a 15" laptop screen with my terminal set to half the width, I can view up to 99 characters with the default iTerm2 settings. Even with the "type" column available, we're less than 80 for a WordPress project with a project name of 9 letters and a ~/git parent folder. Knowing that some may have longer repo name or enclosing folder, that will eat into the remaining ~ 20 characters (although if we drop http://, we'd get back 7). I realize this isn't accounting for those using a 13" or 11" screen, but that should still be doable if the terminal wasn't constrained to half the horizontal width.
- Speculating, but given that Drupal/WordPress crossover isn't super common, I'm not sure noting the type is as valuable for most users. That said, having to dive into `describe` to compare multiple projects would be annoying. I say we leave in for now until we see what feedback we get.

Summarizing:

* I'd prefer 2 spaces between columns for consistency and to reduce horizontal sprawl.
* I think dropping http:// would save space without introducing confusion.
* I think leaving "TYPE" in for now is valuable for now and if we need to shrink space in the future, we can revisit based on feedback.

**Comment by beeradb** [comment]
Just pushed an update with the tabs -> spaces fix. It looks to be a nice improvement. Great suggestion.

As per removing "http://", I experimented with this change and I don't actually think we want to make it. It works fine w/ port 80, but as soon as you add a custom port to the url, save :8080, it no longer opens properly without the `http://`. Since we have plans to offer non-default ports for sites, I would suggest we keep the prefix in place.

**Comment by rickmanelius** [comment]
@beeradb Fair point on http://. I hadn't tested that and that makes sense. Let's leave it.

**Comment by beeradb** [comment]
Waiting for green and then I'll push the button 🎉

**Comment by rfay** [comment]
Awesome, I'll take it! Thanks!
6__default__bash__and_5__default__ddev_
